base_custom_filter/models/base.py

```diff
@@ -54,6 +54,26 @@ def _add_grouped_groupby(self, res, custom_groupbys):
         res["arch"] = etree.tostring(arch)
         return res
 
+    @api.model
+    def _add_search_field(self, res, search_fields):
+        xml_arch = etree.fromstring(res["arch"])
+        for search_field in search_fields:
+            group_ids = self.env.user.groups_id.ids
+            search_field_group_ids = search_field.group_ids.ids
+            if any(group in group_ids for group in search_field_group_ids) or not search_field_group_ids:
```

Comment:
Create a method for this part.
```python
    @api.model
    def _user_has_access_to_item(self, item):
        if not item.group_ids:
            return True
        return bool(set(self.env.user.groups_id.ids) & set(item.group_ids.ids))
```

And then call it like this.

```suggestion
            if not self._user_has_access_to_item(search_field):
                continue
```

Apply this to relevant parts of other methods.